src/haz3lcore/tiles/Segment.re

```diff
@@ -754,3 +754,23 @@ let rec get_incomplete_ids = (seg: t): list(Id.t) =>
 
 let ids_of_incomplete_tiles_in_bidelimiteds = (seg: t): list(Id.t) =>
   get_childrens(seg) |> List.concat |> get_incomplete_ids;
+
+let push_right = ((ls: t, rs: t)): (t, t) =>
+  switch (ls |> List.rev) {
+  | [l, ...ls] => (ls |> List.rev, [l, ...rs])
```

Comment:
It's probably worth adding an `unsnoc` to ListUtil:

```reason
let rec unsnoc = (xs: list('a)): option(('a, list('a))) => {
  switch ((xs: list('a))) {
  | [] => None
  | [head, ...tail] =>
    switch (unsnoc(tail)) {
    | None => Some((head, []))
    | Some((tail_last: 'a, tail_init: list('a))) =>
      Some((tail_last, [head, ...tail_init]))
    }
  };
};
```

Reply:
will do. wish pattern matching just supported this like in civilized languages...

Reply:
:popcorn: :popcorn: 

I get the vibe they don't want you to do it for performance reasons. We could consider an immutable vector or an immutable queue for one side if it becomes an issue.